File: outreach/outreach.py
#!/usr/bin/env python3
"""
StarReviews — Prospection automatique par email
Trouve restaurants/hôtels dans la région Aix-Marseille via OpenStreetMap (100% gratuit).

Usage:
  python outreach.py search          # Trouve les prospects et les sauvegarde dans prospects.csv
  python outreach.py send --dry-run  # Simule l'envoi (sans vraiment envoyer)
  python outreach.py send            # Envoie les emails (max 50/jour par défaut)
  python outreach.py send --limit 30 # Limite à 30 emails
"""

import os
import re
import csv
import time
import smtplib
import argparse
import requests
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart

# ─── Config ───────────────────────────────────────────────────────────────────
GMAIL_USER         = 'starreviewsapp@gmail.com'
GMAIL_APP_PASSWORD = os.environ.get('GMAIL_APP_PASSWORD', '')

# Zones et types de lieux à chercher (OpenStreetMap)
SEARCHES = [
    # (type OSM, ville, rayon en mètres)
    ('restaurant',  'Marseille',         15000),
    ('restaurant',  'Aix-en-Provence',    8000),
    ('restaurant',  'Aubagne',            5000),
    ('restaurant',  'Cassis',             3000),
    ('restaurant',  'La Ciotat',          4000),
    ('restaurant',  'Martigues',          5000),
    ('restaurant',  'Salon-de-Provence',  5000),
    ('hotel',       'Marseille',         15000),
    ('hotel',       'Aix-en-Provence',    8000),
    ('hotel',       'Aubagne',            5000),
    ('hotel',       'Cassis',             3000),
    ('spa',         'Marseille',         15000),
    ('spa',         'Aix-en-Provence',    8000),
]

EMAIL_SUBJECT = "Vos avis Google — j'ai quelque chose pour vous"

EMAIL_BODY = """\
Bonjour {name},

Je m'appelle Clément, j'ai créé StarReviews — un outil qui répond automatiquement à vos avis Google avec l'IA.

Plus besoin de passer du temps chaque semaine à rédiger des réponses. StarReviews s'en charge pour vous, dans le ton que vous choisissez, en quelques secondes.

→ 14 jours gratuits, sans carte bancaire
→ starreviews.vercel.app

Ça vous intéresse ? Je peux vous faire une démo rapide.

Bonne journée,
Clément — StarReviews
starreviewsapp@gmail.com
"""

BLOCKED_EMAIL_DOMAINS = {
    'example.com', 'wordpress.com', 'sentry.io', 'jquery.com',
    'schema.org', 'w3.org', 'wixpress.com', 'squarespace.com',
    'cloudflare.com', 'google.com', 'facebook.com', 'instagram.com',
    'wix.com', 'jimdo.com',
}

# ─── OpenStreetMap (Overpass API) — 100% gratuit ──────────────────────────────
def geocode_city(city: str) -> tuple[float, float] | None:
    """Retourne (lat, lon) d'une ville via Nominatim."""
    url = 'https://nominatim.openstreetmap.org/search'
    params = {'q': city + ', France', 'format': 'json', 'limit': 1}
    headers = {'User-Agent': 'StarReviews-outreach/1.0 (starreviewsapp@gmail.com)'}
    resp = requests.get(url, params=params, headers=headers, timeout=10)
    results = resp.json()
    if results:
        return float(results[0]['lat']), float(results[0]['lon'])
    return None


def search_osm(amenity: str, lat: float, lon: float, radius: int) -> list[dict]:
    """Cherche des lieux via Overpass API (OpenStreetMap)."""
    query = f"""
    [out:json][timeout:25];
    (
      node["amenity"="{amenity}"](around:{radius},{lat},{lon});
      way["amenity"="{amenity}"](around:{radius},{lat},{lon});
      node["tourism"="hotel"](around:{radius},{lat},{lon});
      way["tourism"="hotel"](around:{radius},{lat},{lon});
    );
    out tags center;
    """
    # Pour les hôtels et spas on utilise tourism=hotel / leisure=spa
    if amenity == 'hotel':
        query = f"""
        [out:json][timeout:25];
        (
          node["tourism"="hotel"](around:{radius},{lat},{lon});
          way["tourism"="hotel"](around:{radius},{lat},{lon});
          node["tourism"="motel"](around:{radius},{lat},{lon});
        );
        out tags center;
        """
    elif amenity == 'spa':
        query = f"""
        [out:json][timeout:25];
        (
          node["leisure"="spa"](around:{radius},{lat},{lon});
          way["leisure"="spa"](around:{radius},{lat},{lon});
          node["amenity"="spa"](around:{radius},{lat},{lon});
        );
        out tags center;
        """

    resp = requests.post(
        'https://overpass-api.de/api/interpreter',
        data={'data': query},
        timeout=30,
    )
    elements = resp.json().get('elements', [])

    results = []
    for el in elements:
        tags = el.get('tags', {})
        name = tags.get('name', '')
        if not name:
            continue
        results.append({
            'name':    name,
            'website': tags.get('website', tags.get('contact:website', '')),
            'email':   tags.get('email', tags.get('contact:email', '')),
            'phone':   tags.get('phone', tags.get('contact:phone', '')),
            'address': f"{tags.get('addr:housenumber', '')} {tags.get('addr:street', '')} {tags.get('addr:city', '')}".strip(),
        })
    return results


# ─── Email extraction ──────────────────────────────────────────────────────────
def extract_email_from_website(url: str) -> str | None:
    try:
        headers = {'User-Agent': 'Mozilla/5.0 (compatible; StarReviews-bot/1.0)'}
        resp = requests.get(url, headers=headers, timeout=8, allow_redirects=True)
        emails = re.findall(r'[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}', resp.text)
        emails = [
            e.lower() for e in emails
            if not any(b in e.lower() for b in BLOCKED_EMAIL_DOMAINS)
        ]
        return emails[0] if emails else None
    except Exception:
        return None


# ─── Search mode ───────────────────────────────────────────────────────────────
def run_search(output_file: str = 'prospects.csv'):
    seen: set = set()
    rows: list = []

    city_cache: dict = {}

    for (amenity, city, radius) in SEARCHES:
        print(f'\n🔍  {amenity} — {city}')

        if city not in city_cache:
            coords = geocode_city(city)
            if not coords:
                print(f'  ⚠️  Ville introuvable: {city}')
                continue
            city_cache[city] = coords
            time.sleep(1)

        lat, lon = city_cache[city]
        places = search_osm(amenity, lat, lon, radius)
        time.sleep(1)

        for place in places:
            key = place['name'].lower().strip()
            if key in seen:
                continue
            seen.add(key)

            email = place['email']

            if not email and place['website']:
                email = extract_email_from_website(place['website'])
                time.sleep(0.5)

            rows.append({
                'name':    place['name'],
                'email':   email or '',
                'website': place['website'],
                'phone':   place['phone'],
                'address': place['address'],
                'sent':    'non',
            })
            status = '✅' if email else '❌'
            print(f"  {status}  {place['name']} — {email or 'pas d\\'email'}")

    with open(output_file, 'w', newline='', encoding='utf-8') as f:
        writer = csv.DictWriter(f, fieldnames=['name', 'email', 'website', 'phone', 'address', 'sent'])
        writer.writeheader()
        writer.writerows(rows)

    with_email = sum(1 for r in rows if r['email'])
    print(f'\n📄  {len(rows)} établissements trouvés, {with_email} avec email')
    print(f'    Sauvegardé dans {output_file}')


# ─── Send mode ─────────────────────────────────────────────────────────────────
def run_send(prospects_file: str = 'prospects.csv', dry_run: bool = False, limit: int = 50):
    if not GMAIL_APP_PASSWORD and not dry_run:
        print('❌  GMAIL_APP_PASSWORD manquant dans le fichier .env')
        return

    with open(prospects_file, 'r', encoding='utf-8') as f:
        rows = list(csv.DictReader(f))

    to_send = [r for r in rows if r['email'] and r['sent'] == 'non'][:limit]
    print(f'📧  {len(to_send)} emails à envoyer{"  (DRY RUN — rien ne sera envoyé)" if dry_run else ""}')

    smtp = None
    if not dry_run:
        smtp = smtplib.SMTP_SSL('smtp.gmail.com', 465)
        smtp.login(GMAIL_USER, GMAIL_APP_PASSWORD)

    sent_count = 0
    for row in to_send:
        name  = row['name']
        email = row['email']
        body  = EMAIL_BODY.format(name=name)

        msg = MIMEMultipart('alternative')
        msg['Subject'] = EMAIL_SUBJECT
        msg['From']    = f'Clément — StarReviews <{GMAIL_USER}>'
        msg['To']      = email
        msg.attach(MIMEText(body, 'plain', 'utf-8'))

        if dry_run:
            print(f'  [DRY RUN]  → {name} <{email}>')
        else:
            try:
                smtp.sendmail(GMAIL_USER, email, msg.as_string())
                row['sent'] = 'oui'
                sent_count += 1
                print(f'  ✅  Envoyé → {name} <{email}>')
                time.sleep(3)
            except Exception as e:
                print(f'  ❌  Erreur → {name}: {e}')

    if smtp:
        smtp.quit()

    if not dry_run:
        with open(prospects_file, 'w', newline='', encoding='utf-8') as f:
            writer = csv.DictWriter(f, fieldnames=['name', 'email', 'website', 'phone', 'address', 'sent'])
            writer.writeheader()
            writer.writerows(rows)
        print(f'\n✅  {sent_count} emails envoyés')


# ─── Entry point ───────────────────────────────────────────────────────────────
if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='StarReviews Outreach')
    parser.add_argument('mode', choices=['search', 'send'])
    parser.add_argument('--dry-run', action='store_true')
    parser.add_argument('--limit',   type=int, default=50)
    parser.add_argument('--file',    default='prospects.csv')
    args = parser.parse_args()

    if args.mode == 'search':
        run_search(args.file)
    elif args.mode == 'send':
        run_send(args.file, dry_run=args.dry_run, limit=args.limit)

File: outreach/outreach.mjs
#!/usr/bin/env node
/**
 * StarReviews — Prospection automatique par email
 *
 * Usage:
 *   node outreach.mjs search          → trouve les prospects, sauvegarde dans prospects.csv
 *   node outreach.mjs send --dry-run  → simule sans envoyer
 *   node outreach.mjs send            → envoie les emails (max 50/jour)
 *   node outreach.mjs send --limit 30 → limite à 30 emails
 */

import fs from 'fs'
import path from 'path'
import { fileURLToPath } from 'url'

const __dirname = path.dirname(fileURLToPath(import.meta.url))

// ─── Config ───────────────────────────────────────────────────────────────────
// Charge .env ou .env.local
for (const name of ['.env', '.env.local']) {
  const envPath = path.join(__dirname, name)
  if (fs.existsSync(envPath)) {
    for (const line of fs.readFileSync(envPath, 'utf8').split('\n')) {
      const [k, ...v] = line.split('=')
      if (k && v.length) process.env[k.trim()] = v.join('=').trim()
    }
  }
}

const BREVO_API_KEY = process.env.BREVO_API_KEY || ''
const SENDER_EMAIL  = 'starreviewsapp@gmail.com'
const SENDER_NAME   = 'Clément — StarReviews'

// Coordonnées hardcodées pour éviter les rate limits de Nominatim
const CITY_COORDS = {
  'Marseille':         [43.2965, 5.3698],
  'Aix-en-Provence':  [43.5297, 5.4474],
  'Aubagne':           [43.2967, 5.5709],
  'Cassis':            [43.2140, 5.5389],
  'La Ciotat':         [43.1742, 5.6083],
  'Martigues':         [43.4054, 5.0471],
  'Salon-de-Provence': [43.6408, 5.0979],
}

const SEARCHES = [
  ['restaurant', 'Marseille',        15000],
  ['restaurant', 'Aix-en-Provence',   8000],
  ['restaurant', 'Aubagne',           5000],
  ['restaurant', 'Cassis',            3000],
  ['restaurant', 'La Ciotat',         4000],
  ['restaurant', 'Martigues',         5000],
  ['restaurant', 'Salon-de-Provence', 5000],
  ['hotel',      'Marseille',        15000],
  ['hotel',      'Aix-en-Provence',   8000],
  ['hotel',      'Aubagne',           5000],
  ['hotel',      'Cassis',            3000],
  ['spa',        'Marseille',        15000],
  ['spa',        'Aix-en-Provence',   8000],
]

const EMAIL_SUBJECT = "Vos avis Google — j'ai quelque chose pour vous"

const EMAIL_BODY = (name) => `Bonjour ${name},

Je m'appelle Clément, j'ai créé StarReviews — un outil qui répond automatiquement à vos avis Google avec l'IA.

Plus besoin de passer du temps chaque semaine à rédiger des réponses. StarReviews s'en charge pour vous, dans le ton que vous choisissez, en quelques secondes.

→ 14 jours gratuits, sans carte bancaire
→ starreviews.vercel.app

Ça vous intéresse ? Je peux vous faire une démo rapide.

Bonne journée,
Clément — StarReviews
starreviewsapp@gmail.com
`

const BLOCKED_DOMAINS = ['example.com','wordpress.com','sentry.io','schema.org',
  'w3.org','cloudflare.com','google.com','facebook.com','instagram.com','wix.com']

// ─── Helpers ──────────────────────────────────────────────────────────────────
const sleep = (ms) => new Promise(r => setTimeout(r, ms))

async function fetchJson(url, opts = {}) {
  const res = await fetch(url, { headers: { 'User-Agent': 'StarReviews-outreach/1.0 (starreviewsapp@gmail.com)' }, ...opts })
  const text = await res.text()
  try { return JSON.parse(text) } catch { return null }
}

async function geocodeCity(city) {
  const data = await fetchJson(`https://nominatim.openstreetmap.org/search?q=${encodeURIComponent(city + ', France')}&format=json&limit=1`)
  if (Array.isArray(data) && data[0]) return [parseFloat(data[0].lat), parseFloat(data[0].lon)]
  return null
}

async function searchOSM(amenity, lat, lon, radius, retry = 0) {
  let query
  if (amenity === 'hotel') {
    query = `[out:json][timeout:25];(node["tourism"="hotel"](around:${radius},${lat},${lon});way["tourism"="hotel"](around:${radius},${lat},${lon}););out tags center;`
  } else if (amenity === 'spa') {
    query = `[out:json][timeout:25];(node["leisure"="spa"](around:${radius},${lat},${lon});way["leisure"="spa"](around:${radius},${lat},${lon}););out tags center;`
  } else {
    query = `[out:json][timeout:25];(node["amenity"="${amenity}"](around:${radius},${lat},${lon});way["amenity"="${amenity}"](around:${radius},${lat},${lon}););out tags center;`
  }
  const servers = ['https://overpass-api.de/api/interpreter', 'https://overpass.kumi.systems/api/interpreter']
  const url = servers[retry % servers.length]
  let text
  try {
    const res = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        'User-Agent': 'StarReviews-outreach/1.0 (starreviewsapp@gmail.com)',
      },
      body: new URLSearchParams({ data: query }),
      signal: AbortSignal.timeout(30000),
    })
    text = await res.text()
  } catch {
    if (retry < 2) { await sleep(3000); return searchOSM(amenity, lat, lon, radius, retry + 1) }
    return []
  }
  let data
  try { data = JSON.parse(text) } catch {
    console.log('  ⚠️  Réponse API non-JSON:', text.slice(0, 200))
    if (retry < 2) { await sleep(3000); return searchOSM(amenity, lat, lon, radius, retry + 1) }
    return []
  }
  console.log(`  ℹ️  ${data.elements?.length ?? 0} éléments bruts reçus`)
  return (data.elements || []).map(el => {
    const t = el.tags || {}
    return {
      name:    t.name || '',
      website: t.website || t['contact:website'] || '',
      email:   t.email || t['contact:email'] || '',
      phone:   t.phone || t['contact:phone'] || '',
      address: [t['addr:housenumber'], t['addr:street'], t['addr:city']].filter(Boolean).join(' '),
    }
  }).filter(p => p.name)
}

async function extractEmailFromWebsite(siteUrl) {
  const pagesToTry = [siteUrl, siteUrl.replace(/\/?$/, '/contact'), siteUrl.replace(/\/?$/, '/nous-contacter')]
  for (const url of pagesToTry) {
    try {
      const res = await fetch(url, {
        headers: { 'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36' },
        signal: AbortSignal.timeout(6000),
      })
      const html = await res.text()
      const matches = html.match(/[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}/g) || []
      const filtered = matches.filter(e => !BLOCKED_DOMAINS.some(d => e.toLowerCase().includes(d)))
      if (filtered[0]) return filtered[0].toLowerCase()
    } catch { /* continue */ }
  }
  return null
}

// ─── CSV ──────────────────────────────────────────────────────────────────────
function readCsv(file) {
  if (!fs.existsSync(file)) return []
  const lines = fs.readFileSync(file, 'utf8').trim().split('\n')
  const headers = lines[0].split(',')
  return lines.slice(1).map(line => {
    const vals = line.split(',')
    return Object.fromEntries(headers.map((h, i) => [h, vals[i] || '']))
  })
}

function writeCsv(file, rows) {
  const headers = ['name','email','website','phone','address','sent']
  const lines = [headers.join(','), ...rows.map(r => headers.map(h => (r[h] || '').replace(/,/g, ' ')).join(','))]
  fs.writeFileSync(file, lines.join('\n'), 'utf8')
}

// ─── Search ───────────────────────────────────────────────────────────────────
async function runSearch(outputFile = 'prospects.csv') {
  const seen = new Set()
  const rows = []
  const cityCache = {}

  for (const [amenity, city, radius] of SEARCHES) {
    console.log(`\n🔍  ${amenity} — ${city}`)

    if (!cityCache[city]) {
      // Utilise les coordonnées hardcodées, sinon Nominatim
      if (CITY_COORDS[city]) {
        cityCache[city] = CITY_COORDS[city]
      } else {
        const coords = await geocodeCity(city)
        if (!coords) { console.log(`  ⚠️  Ville introuvable: ${city}`); continue }
        cityCache[city] = coords
        await sleep(1000)
      }
    }

    const [lat, lon] = cityCache[city]
    const places = await searchOSM(amenity, lat, lon, radius)
    await sleep(1000)

    for (const place of places) {
      const key = place.name.toLowerCase().trim()
      if (seen.has(key)) continue
      seen.add(key)

      let email = place.email
      if (!email && place.website) {
        email = await extractEmailFromWebsite(place.website)
        await sleep(500)
      }

      rows.push({ ...place, email: email || '', sent: 'non' })
      console.log(`  ${email ? '✅' : '❌'}  ${place.name} — ${email || 'pas d\'email'}`)
    }
  }

  writeCsv(outputFile, rows)
  const withEmail = rows.filter(r => r.email).length
  console.log(`\n📄  ${rows.length} établissements trouvés, ${withEmail} avec email`)
  console.log(`    Sauvegardé dans ${outputFile}`)
}

// ─── Send ─────────────────────────────────────────────────────────────────────
async function runSend(prospectsFile = 'prospects.csv', dryRun = false, limit = 50) {
  if (!BREVO_API_KEY && !dryRun) {
    console.log('❌  BREVO_API_KEY manquant dans le fichier .env')
    return
  }

  const rows = readCsv(prospectsFile)
  const toSend = rows.filter(r => r.email && r.sent === 'non').slice(0, limit)
  console.log(`📧  ${toSend.length} emails à envoyer${dryRun ? '  (DRY RUN)' : ''}`)

  let sentCount = 0
  for (const row of toSend) {
    if (dryRun) {
      console.log(`  [DRY RUN]  → ${row.name} <${row.email}>`)
      continue
    }
    const res = await fetch('https://api.brevo.com/v3/smtp/email', {
      method: 'POST',
      headers: { 'api-key': BREVO_API_KEY, 'Content-Type': 'application/json' },
      body: JSON.stringify({
        sender:      { name: SENDER_NAME, email: SENDER_EMAIL },
        to:          [{ email: row.email, name: row.name }],
        subject:     EMAIL_SUBJECT,
        textContent: EMAIL_BODY(row.name),
      }),
    })
    if (res.ok) {
      row.sent = 'oui'
      sentCount++
      console.log(`  ✅  Envoyé → ${row.name} <${row.email}>`)
    } else {
      console.log(`  ❌  Erreur → ${row.name} <${row.email}>`)
    }
    await sleep(1000)
  }

  writeCsv(prospectsFile, rows)
  console.log(`\n✅  ${sentCount} emails envoyés`)
}

// ─── Auto mode (search + send en boucle toutes les 23h) ──────────────────────
async function runAuto(prospectsFile = 'prospects.csv', dailyLimit = 10) {
  let round = 1
  while (true) {
    console.log(`\n${'─'.repeat(50)}`)
    console.log(`🔄  Tour ${round} — ${new Date().toLocaleString('fr-FR')}`)
    console.log('─'.repeat(50))

    await runSearch(prospectsFile)
    await runSend(prospectsFile, false, dailyLimit)

    const nextRun = new Date(Date.now() + 23 * 60 * 60 * 1000)
    console.log(`\n⏰  Prochain envoi dans 23h (${nextRun.toLocaleString('fr-FR')})`)
    console.log('   Laisse cette fenêtre ouverte.')
    round++
    await sleep(23 * 60 * 60 * 1000)
  }
}

// ─── Main ─────────────────────────────────────────────────────────────────────
const [,, mode, ...flags] = process.argv
const dryRun = flags.includes('--dry-run')
const limitIdx = flags.indexOf('--limit')
const limit = limitIdx >= 0 ? parseInt(flags[limitIdx + 1]) : 10
const fileIdx = flags.indexOf('--file')
const file = fileIdx >= 0 ? flags[fileIdx + 1] : 'prospects.csv'

if (mode === 'search') {
  runSearch(file)
} else if (mode === 'send') {
  runSend(file, dryRun, limit)
} else if (mode === 'auto') {
  runAuto(file, limit)
} else {
  console.log('Usage: node outreach.mjs search | send | auto [--limit N]')
  console.log('  auto  → cherche + envoie automatiquement toutes les 23h')
}
